### PR TITLE
Corrected shortcut

### LINKED ISSUE #6867: Double key binding for inserting new TechReport and pulling changes from shared database
JabRef version 5.2 on openSUSE Leap 15.1

- [x] **Mandatory**: I have tested the latest development version from http://builds.jabref.org/master/ and the problem persists

1. Open a shared database.
2. Press "Ctrl+Shift+R".
3. Changes from shared database are pulled **and** a new TechReport is inserted.

Maybe this is quickly done... ;)

### PR DESCRIPTION
**Pull request by Gena928**
Fixes #6867 
A new shortcut combination added for "Pull changes from shared database", because current ctrl+shift+R is already assigned for "Create tech report" command.

- [X] Change in CHANGELOG.md described (if applicable)
- [ ] Tests created for changes (if applicable)
- [X] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, submitted a pull request to the documentation repository.

### CODE REVIEW

**Comment by tobiasdiez** [comment]
I would remove all the standard keybindings for https://github.com/JabRef/jabref/pull/6960/files#diff-920a3cf78140f82ffcd3ba5b495cf73fL49-L54. Then you can leave the pull changes keybinding at "ctrl + shift + r". What do you think?

**Comment by Siedlerchr** [comment]
I'm against removing the shortcuts, at least article should be kept. Let's say what the other @JabRef/developers say about this

**Comment by tobiasdiez** [comment]
Yes, article, book, new entry and new entry from plaintext should be kept. But masterthesis, phdthesis etc are not that important to have a default keybinding. (I don't suggest to completely remove these bindings, just their default value so that users can still activate them if they want).

**Comment by Gena928** [comment]
@tobiasdiez ,
So the following "new" shortcuts will be removed (see gray color):
Screenshot 2020-10-05 at 10 51 53

correct?

**Comment by calixtus** [comment]
I wouldn't remove them at all, but their default value.
So the user should be able, to bind keys to the shortcuts, if he likes.

**Comment by tobiasdiez** [comment]
Yes, just provide an empty string as the third argument of these enum values, and include NEW_TECHREPORT as well.

**Comment by Gena928** [comment]
Empty string doesn't works. )) When I try to open the menu, it shows an error. 
I'll try to find the problem.

java.lang.StringIndexOutOfBoundsException: begin 0, end 1, length 0
	at java.base/java.lang.String.checkBoundsBeginEnd(String.java:3756)
	at java.base/java.lang.String.substring(String.java:1902)
	at javafx.graphics/javafx.scene.input.KeyCombination.normalizeToken(KeyCombination.java:630)
	at javafx.graphics/javafx.scene.input.KeyCombination.valueOf(KeyCombination.java:409)
	at org.jabref@100.0.0/org.jabref.gui.keyboard.KeyBindingRepository.getKeyCombination(KeyBindingRepository.java:129)
	at org.jabref@100.0.0/org.jabref.gui.keyboard.KeyBindingRepository.checkKeyCombinationEquality(KeyBindingRepository.java:140)
	at org.jabref@100.0.0/org.jabref.gui.keyboard.KeyBindingRepository.mapToKeyBinding(KeyBindingRepository.java:116)
	at org.jabref@100.0.0/org.jabref.gui.JabRefFrame.lambda$initKeyBindings$5(JabRefFrame.java:236)
	at javafx.base/com.sun.javafx.event.CompositeEventHandler$NormalEventFilterRecord.handleCapturingEvent(CompositeEventHandler.java:321)
	at javafx.base/com.sun.javafx.event.CompositeEventHandler.dispatchCapturingEvent(CompositeEventHandler.java:98)
	at javafx.base/com.sun.javafx.event.EventHandlerManager.dispatchCapturingEvent(EventHandlerManager.java:221)
	at javafx.base/com.sun.javafx.event.EventHandlerManager.dispatchCapturingEvent(EventHandlerManager.java:180)
	at javafx.base/com.sun.javafx.event.CompositeEventDispatcher.dispatchCapturingEvent(CompositeEventDispatcher.java:43)
	at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:52)
	at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at javafx.base/com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
	at javafx.base/com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:54)
	at javafx.base/javafx.event.Event.fireEvent(Event.java:198)
	at javafx.graphics/javafx.scene.Scene$KeyHandler.process(Scene.java:4098)
	at javafx.graphics/javafx.scene.Scene.processKeyEvent(Scene.java:2157)
	at javafx.graphics/javafx.scene.Scene$ScenePeerListener.keyEvent(Scene.java:2625)
	at javafx.graphics/com.sun.javafx.tk.quantum.GlassViewEventHandler$KeyEventNotification.run(GlassViewEventHandler.java:217)
	at javafx.graphics/com.sun.javafx.tk.quantum.GlassViewEventHandler$KeyEventNotification.run(GlassViewEventHandler.java:149)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:391)
	at javafx.graphics/com.sun.javafx.tk.quantum.GlassViewEventHandler.lambda$handleKeyEvent$1(GlassViewEventHandler.java:248)
	at javafx.graphics/com.sun.javafx.tk.quantum.QuantumToolkit.runWithoutRenderLock(QuantumToolkit.java:412)
	at javafx.graphics/com.sun.javafx.tk.quantum.GlassViewEventHandler.handleKeyEvent(GlassViewEventHandler.java:247)
	at javafx.graphics/com.sun.glass.ui.View.handleKeyEvent(View.java:547)
	at javafx.graphics/com.sun.glass.ui.View.notifyKey(View.java:971)

**Comment by koppor** [comment]
@Gena928 You pasated a PNG instead of using a link into the code (offered by the GitHub UI) (or maybe additoinally)

The affected keys are at `org.jabref.gui.keyboard.KeyBinding#NEW_MASTERSTHESIS`.

Can't we just add `alt+` at them to a) provide a shortcut (seems to be required according to the comment) and b) having a default short cut surely not conflicting?

@Gena928 While you are on it, could you please add a shortcot for `New inproceedings`? In computer science we use that type in 80% of our cases and that would make JabRef more comfortable to use.

**Comment by Gena928** [comment]
@koppor 
I'm very sorry for delay with this Issue. I'll try to fix it today, including shortcut for `New inprocessing`.

**Comment by tobiasdiez** [comment]
@Gena928 You should be able to fix this error by adding a few additional if-statements if the keybinding string is empty.

**Comment by Gena928** [comment]
Please review changes.
1) I was solved the problem with empty `defaultKeyBinding` string by adding an `if` statement in `KeyBindingRepository.java`.
2) Key combinations for NEW_INBOOK...NEW_TECHREPORT changed from `ctrl+....` to `alt+....`
3) I think now we can keep `ctrl+shift+R` for PULL_CHANGES_FROM_SHARED_DATABASE.
As it was initially. What do you think?

4) I also created NEW_INPROCEEDINGS KeyBinding.
But I cant't find this menu item in `createMenu()` method of `JabRefFrame` class. ))) Sorry, need help. Do we have such menu item? Where? )))

**Comment by koppor** [comment]
> @tobiasdiez , @koppor Please review changes.

Could you please resolve the conflicts. Please merge `upstream/master`.

>     1. I solved the problem with empty `defaultKeyBinding` string by adding an `if` statement in `KeyBindingRepository.java`.

LGTM.

>     2. Key combinations for NEW_INBOOK...NEW_TECHREPORT changed from `ctrl+....` to `alt+....`

It was on purpose to **add** an additional modifier and not to change the modifier. I was thinking that an additoinal moddifier is easer to remember instead of another modifier.

However, having our discussion in mind, we should just leave these shortcuts empty. A user seldomly will use these shortcuts.
 
>     3. I think now we can keep `ctrl+shift+R` for PULL_CHANGES_FROM_SHARED_DATABASE.
>        As it was initially. What do you think?

That's good. Please revert your change to change the modifier to `alt+shift+D`.

>     4. I also created NEW_INPROCEEDINGS KeyBinding.

Looks good.

>        But I cant't find this menu item in `createMenu()` method of `JabRefFrame` class. ))) Sorry, need help. Do we have such menu item? Where? )))

I don't quite get what you are referring to with "this menu item". In the normal entry creation menu, it is already there: https://docs.jabref.org/collect/add-entry-manually  I think, your current code enough.

I think, it should also be added to `org.jabref.gui.fieldeditors.TypeEditorViewModel#TypeEditorViewModel`

**Comment by Gena928** [comment]
@koppor 
Done. Sorry, for "menu item". I thought you can attach key bindings only to menu items, but now I know about keys for  [https://docs.jabref.org/collect/add-entry-manually](https://docs.jabref.org/collect/add-entry-manually). New InProceedings works via shortcut `alt+shift+C`.

Please review `org.jabref.gui.fieldeditors.TypeEditorViewModel`. I added new line there, but I still can't understand where and how it's used. ))) Can you give a small comment?

File: src/main/java/org/jabref/gui/fieldeditors/TypeEditorViewModel.java [hunk 1]
```diff
@@ -23,6 +23,7 @@ public TypeEditorViewModel(Field field, SuggestionProvider<?> suggestionProvider
         itemMap.put("software", Localization.lang("Software"));
         itemMap.put("datacd", Localization.lang("Data CD"));
         itemMap.put("audiocd", Localization.lang("Audio CD"));
+        itemMap.put("inproceedings", Localization.lang("InProceedings"));
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/fieldeditors/TypeEditorViewModel.java:26]: You don't need to add it here. This class is only used for the field `type` in the entry editor, and there are predefined values for this field.

File: src/main/java/org/jabref/gui/keyboard/KeyBinding.java [hunk 2]
```diff
@@ -46,12 +46,13 @@
...
-    NEW_PHDTHESIS("New phdthesis", Localization.lang("New phdthesis"), "ctrl+shift+T", KeyBindingCategory.BIBTEX),
-    NEW_PROCEEDINGS("New proceedings", Localization.lang("New proceedings"), "ctrl+shift+P", KeyBindingCategory.BIBTEX),
-    NEW_UNPUBLISHED("New unpublished", Localization.lang("New unpublished"), "ctrl+shift+U", KeyBindingCategory.BIBTEX),
-    NEW_TECHREPORT("New technical report", Localization.lang("New technical report"), "ctrl+shift+R", KeyBindingCategory.BIBTEX),
+    NEW_INBOOK("New inbook", Localization.lang("New inbook"), "", KeyBindingCategory.BIBTEX),
+    NEW_MASTERSTHESIS("New mastersthesis", Localization.lang("New mastersthesis"), "", KeyBindingCategory.BIBTEX),
+    NEW_PHDTHESIS("New phdthesis", Localization.lang("New phdthesis"), "", KeyBindingCategory.BIBTEX),
+    NEW_PROCEEDINGS("New proceedings", Localization.lang("New proceedings"), "", KeyBindingCategory.BIBTEX),
+    NEW_UNPUBLISHED("New unpublished", Localization.lang("New unpublished"), "", KeyBindingCategory.BIBTEX),
+    NEW_TECHREPORT("New technical report", Localization.lang("New technical report"), "alt+shift+R", KeyBindingCategory.BIBTEX),
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/keyboard/KeyBinding.java:54]: Please remove the default shortcut for techreport and inprocceedings as well.

File: src/main/java/org/jabref/gui/keyboard/KeyBinding.java [hunk 3]
```diff
@@ -61,7 +62,7 @@
     OPEN_OPEN_OFFICE_LIBRE_OFFICE_CONNECTION("Open OpenOffice/LibreOffice connection", Localization.lang("Open OpenOffice/LibreOffice connection"), "alt+0", KeyBindingCategory.TOOLS),
     OPEN_URL_OR_DOI("Open URL or DOI", Localization.lang("Open URL or DOI"), "F3", KeyBindingCategory.TOOLS),
     PASTE("Paste", Localization.lang("Paste"), "ctrl+V", KeyBindingCategory.EDIT),
-    PULL_CHANGES_FROM_SHARED_DATABASE("Pull changes from shared database", Localization.lang("Pull changes from shared database"), "ctrl+shift+R", KeyBindingCategory.FILE),
+    PULL_CHANGES_FROM_SHARED_DATABASE("Pull changes from shared database", Localization.lang("Pull changes from shared database"), "alt+shift+D", KeyBindingCategory.FILE),
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/keyboard/KeyBinding.java:65]: Please revert to ctrl + shrift + r

File: src/main/java/org/jabref/gui/keyboard/KeyBindingRepository.java [hunk 4]
```diff
@@ -113,8 +113,10 @@ public int size() {
 
     public Optional<KeyBinding> mapToKeyBinding(KeyEvent keyEvent) {
         for (KeyBinding binding : KeyBinding.values()) {
-            if (checkKeyCombinationEquality(binding, keyEvent)) {
-                return Optional.of(binding);
+            if (!binding.getDefaultKeyBinding().isEmpty()){
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/gui/keyboard/KeyBindingRepository.java:116]: Are you sure this works? I would gess `getDefaultKeyBinding` is always empty for `NEW_INBOOK` even though the user has overwritten it. On first glance, I would suggest to change the `getKeyCombination` method below to `public Optional<KeyCombination> getKeyCombination(KeyBinding bindName)` and return an empty optional if `binding` is empty.

Moreover, `get` should probably return en empty string instead of `"Not associated"` to be consistent.

**Review by tobiasdiez — CHANGES_REQUESTED** [review]

File: src/main/java/org/jabref/gui/fieldeditors/TypeEditorViewModel.java (hunk 1 above)
Comment (koppor) [reply src/main/java/org/jabref/gui/fieldeditors/TypeEditorViewModel.java:26]: Ah, my fault. What code comment should go there to point to the code where this happens? 😇

**Comment by koppor** [comment]
@Gena928 Please try to merge upstream/master to resolve the merge conflicts.

File: src/main/java/org/jabref/gui/keyboard/KeyBinding.java (hunk 2 above)
Comment (Gena928) [reply src/main/java/org/jabref/gui/keyboard/KeyBinding.java:54]: OK

File: src/main/java/org/jabref/gui/keyboard/KeyBinding.java (hunk 3 above)
Comment (Gena928) [reply src/main/java/org/jabref/gui/keyboard/KeyBinding.java:65]: OK

File: src/main/java/org/jabref/gui/keyboard/KeyBindingRepository.java (hunk 4 above)
Comment (Gena928) [reply src/main/java/org/jabref/gui/keyboard/KeyBindingRepository.java:116]: It works. If I change the next statement 
from `public KeyCombination getKeyCombination(KeyBinding bindName)` 
to `Optional<KeyCombination> getKeyCombination(KeyBinding bindName)` 

I have to change other parts of code. For example line 141 
from `KeyCombination keyCombination = getKeyCombination(binding);`
to `KeyCombination keyCombination = getKeyCombination(binding).get()`;

Is this what you mean?

Comment (tobiasdiez) [reply src/main/java/org/jabref/gui/keyboard/KeyBindingRepository.java:116]: Okay, can you briefly explain why it works? I thought "getDefaultKeyBinding" then always returns an empty string, so this check is always false.

Yeah, there are more code changes required for my suggestion. I think it would provide a clean solution to handle the case that for a given keybinding there is not necessarily a keycombination. But I didn't had the time yet to check the code in detail. So if your solution indeed works, then it's fine with me as well.

Comment (Gena928) [reply src/main/java/org/jabref/gui/keyboard/KeyBindingRepository.java:116]: I started debugger and see that it always returns shortcut from `KeyBinding.java` file. 

Screenshot 2020-10-14 at 12 56 37

Solution with `Optional` looks more clean. I just want to confirm that if we implement it, we also change other parts of code to `getKeyCombination(binding).get()`.

Comment (koppor) [reply src/main/java/org/jabref/gui/keyboard/KeyBindingRepository.java:116]: In case this is done, the code fragment is `getKeyCombination(binding).ifpresent(combination -> ...)`

Comment (tobiasdiez) [reply src/main/java/org/jabref/gui/keyboard/KeyBindingRepository.java:116]: @Gena928 Exactly! So if you change the keybinding for "New inbook" to something, then `getDefaultKeybinding` should still return an empty string, so the code never goes into the if body in this case (although it should to match the new-user defined key).

Comment (Gena928) [reply src/main/java/org/jabref/gui/keyboard/KeyBindingRepository.java:116]: @tobiasdiez 
I agree. If you change default shortcut, this code will not work.

Comment (Gena928) [reply src/main/java/org/jabref/gui/keyboard/KeyBindingRepository.java:116]: I'll try to fix today.

**Comment by Gena928** [comment]
@koppor , @tobiasdiez ,
please review changes. Sorry for delay - I'm trying to learn Java with mentor and become valuable contributor )). So thank you for your comments here. They really help.
I changed `getKeyCombination` method of `KeyBindingRepository` class and all it's dependencies.

Also please give your comments for TypeEditorViewModel class - this [comment from @tobiasdiez](https://github.com/JabRef/jabref/pull/6960#discussion_r504478716)

File: src/main/java/org/jabref/gui/keyboard/KeyBindingRepository.java [hunk 5]
```diff
@@ -112,21 +112,23 @@ public int size() {
     }
 
     public Optional<KeyBinding> mapToKeyBinding(KeyEvent keyEvent) {
-        for (KeyBinding binding : KeyBinding.values()) {
+        for (KeyBinding binding : KeyBinding.values()){
```
Comment (calixtus) [inline src/main/java/org/jabref/gui/keyboard/KeyBindingRepository.java:115]: Checkstyle issue: Put the space back between `)` and `{`

File: CHANGELOG.md [hunk 6]
```diff
@@ -45,13 +45,16 @@ inserting new citations in a OpenOffic/LibreOffice document. [#6957](https://git
 - We fixed the failure to Copy citation key and link. [#5835](https://github.com/JabRef/jabref/issues/5835)
 - We fixed an issue where the sort order of the entry table was reset after a restart of JabRef. [#6898](https://github.com/JabRef/jabref/pull/6898)
 - We fixed an issue where no longer a warning was displayed when inserting references into LibreOffice with an invalid "ReferenceParagraphFormat". [#6907](https://github.com/JabRef/jabref/pull/60907).
+- We fixed an issue where a selected field was not removed after the first click in the custom entry types dialog [#6934](https://github.com/JabRef/jabref/issues/6934)
+- We fixed an issue where a remove icon was shown for standard entry types in the custom entry types dialog [6906](https://github.com/JabRef/jabref/issues/6906)
+- We fixed an issue, when pulling changes from shared database via shortcut caused creation a new new tech report [6867](https://github.com/JabRef/jabref/issues/6867)
```
Comment (calixtus) [inline CHANGELOG.md:50]: Merge error?

File: src/main/java/org/jabref/gui/keyboard/KeyBinding.java [hunk 7]
```diff
@@ -46,12 +46,13 @@
...
-    NEW_PROCEEDINGS("New proceedings", Localization.lang("New proceedings"), "ctrl+shift+P", KeyBindingCategory.BIBTEX),
-    NEW_UNPUBLISHED("New unpublished", Localization.lang("New unpublished"), "ctrl+shift+U", KeyBindingCategory.BIBTEX),
-    NEW_TECHREPORT("New technical report", Localization.lang("New technical report"), "ctrl+shift+R", KeyBindingCategory.BIBTEX),
+    NEW_INBOOK("New inbook", Localization.lang("New inbook"), "", KeyBindingCategory.BIBTEX),
+    NEW_MASTERSTHESIS("New mastersthesis", Localization.lang("New mastersthesis"), "", KeyBindingCategory.BIBTEX),
+    NEW_PHDTHESIS("New phdthesis", Localization.lang("New phdthesis"), "", KeyBindingCategory.BIBTEX),
+    NEW_PROCEEDINGS("New proceedings", Localization.lang("New proceedings"), "", KeyBindingCategory.BIBTEX),
+    NEW_UNPUBLISHED("New unpublished", Localization.lang("New unpublished"), "", KeyBindingCategory.BIBTEX),
+    NEW_TECHREPORT("New technical report", Localization.lang("New technical report"), "", KeyBindingCategory.BIBTEX),
+    NEW_INPROCEEDINGS("New inproceesings", Localization.lang("New inproceedings"), "", KeyBindingCategory.BIBTEX),
```
Comment (calixtus) [inline src/main/java/org/jabref/gui/keyboard/KeyBinding.java:55]: `New inproceedings` needs to be put in the english localization file.

**Review by calixtus — CHANGES_REQUESTED** [review]
Good work so far, keep on coding and learning.
Got three little remarks.

**Comment by Gena928** [comment]
@calixtus ,
thanks for comments. Please review changes.

File: CHANGELOG.md [hunk 8]
```diff
@@ -49,9 +49,10 @@ inserting new citations in a OpenOffic/LibreOffice document. [#6957](https://git
 - We fixed an issue where a remove icon was shown for standard entry types in the custom entry types dialog. [#6906](https://github.com/JabRef/jabref/issues/6906)
 - We fixed an issue where it was impossible to connect to OpenOffice/LibreOffice on Mac OSX. [#6970](https://github.com/JabRef/jabref/pull/6970)
 - We fixed an issue with the python script used by browser plugins that failed to locate JabRef if not installed in its default location. [#6963](https://github.com/JabRef/jabref/pull/6963/files)
+- We fixed an issue where spaces and newlines in an isbn would generate an exception. [#6456](https://github.com/JabRef/jabref/issues/6456)
```
Comment (calixtus) [inline CHANGELOG.md:52]: This line still seems to be unrelated to your changes...

File: src/main/resources/l10n/JabRef_en.properties [hunk 9]
```diff
@@ -2257,6 +2257,7 @@ Previous\ preview\ style=Previous preview style
 
 (\ Note\:\ Press\ return\ to\ commit\ changes\ in\ the\ table\!\ )=( Note\: Press return to commit changes in the table\! )
 Reset=Reset
+New\ inproceedings=New article in a conference proceedings
```
Comment (calixtus) [inline src/main/resources/l10n/JabRef_en.properties:2260]: This needs to be exactly identical for the consistency test.
See: https://github.com/JabRef/jabref/runs/1271802665?check_suite_focus=true

```suggestion
New\ inproceedings=New inproceedings
```

File: src/main/java/org/jabref/gui/keyboard/KeyBindingRepository.java [hunk 10]
```diff
@@ -137,8 +139,11 @@ public KeyCombination getKeyCombination(KeyBinding bindName) {
      * @return true if matching, else false
      */
     public boolean checkKeyCombinationEquality(KeyBinding binding, KeyEvent keyEvent) {
-        KeyCombination keyCombination = getKeyCombination(binding);
-        return checkKeyCombinationEquality(keyCombination, keyEvent);
+        Optional<KeyCombination> keyCombination = getKeyCombination(binding);
+        if (!keyCombination.isPresent()){
```
Comment (calixtus) [inline src/main/java/org/jabref/gui/keyboard/KeyBindingRepository.java:143]: Missing whitespace.

```suggestion
        if (!keyCombination.isPresent()) {
```

**Review by calixtus — CHANGES_REQUESTED** [review]
Still three issues. ;-)

**Comment by Gena928** [comment]
@calixtus 
thanks again, corrected.
Looks like my "CheckStyle" was disabled. Maybe update of Intellij?

**Review by calixtus — APPROVED** [review]
LGTM to me now. Thanks for your efforts!

**Review by koppor — APPROVED** [review]

**Comment by koppor** [comment]
Also OK for me. Still have to update CHANGELOG.md, because keyboard shortcuts were removed in this PR.
